File: scripts/quantize-b52-d11-1-motion.py
#!/usr/bin/env python3
"""Bounded scalar-Python motion quantizer for B52-D11.1."""

from __future__ import annotations

import argparse
import hashlib
import json
import math
import os
import struct
import sys
from pathlib import Path


SPEC_SHA256 = "c4cb343672f53660d7c4ab69ccd489e00bb211e4aa1f489429f7a626ee48c42a"


def sha(path: Path) -> str:
    return hashlib.sha256(path.read_bytes()).hexdigest()


def canonical_hash(value: object) -> str:
    payload = json.dumps(value, sort_keys=True, separators=(",", ":"), allow_nan=False).encode()
    return hashlib.sha256(payload).hexdigest()


def arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--spec", type=Path, required=True)
    parser.add_argument("--fixture", required=True)
    parser.add_argument("--repeat", type=int, choices=(1, 2), required=True)
    parser.add_argument("--input", type=Path, required=True)
    parser.add_argument("--adapter-report", type=Path, required=True)
    parser.add_argument("--output", type=Path, required=True)
    parser.add_argument("--report", type=Path, required=True)
    return parser.parse_args()


def nearest_integer(value: float) -> int:
    return math.floor(value + 0.5) if value >= 0.0 else math.ceil(value - 0.5)


def quantize(payload: bytes, radius: float) -> tuple[bytes, float]:
    if len(payload) % 4:
        raise RuntimeError("motion payload is not a whole little-endian float32 array")
    values = struct.unpack(f"<{len(payload) // 4}f", payload)
    output: list[float] = []
    maximum_error = 0.0
    for index, value in enumerate(values):
        if not math.isfinite(value):
            raise RuntimeError(f"QUANTIZER_DOMAIN nonfinite component={index}")
        candidate = nearest_integer(value)
        error = abs(value - candidate)
        if error > radius:
            raise RuntimeError(
                f"QUANTIZER_DOMAIN component={index} value={value!r} candidate={candidate} error={error!r} radius={radius!r}"
            )
        maximum_error = max(maximum_error, error)
        output.append(0.0 if candidate == 0 else float(candidate))
    encoded = struct.pack(f"<{len(output)}f", *output)
    if quantize_integral(encoded) != encoded:
        raise RuntimeError("quantizer idempotence failure")
    return encoded, maximum_error


def quantize_integral(payload: bytes) -> bytes:
    values = struct.unpack(f"<{len(payload) // 4}f", payload)
    output = [0.0 if value == 0.0 else float(nearest_integer(value)) for value in values]
    return struct.pack(f"<{len(output)}f", *output)


def main() -> None:
    args = arguments()
    spec = json.loads(args.spec.read_text())
    fixture = next((item for item in spec["fixtures"] if item["id"] == args.fixture), None)
    if sha(args.spec) != SPEC_SHA256 or fixture is None:
        raise RuntimeError("B52-D11.1 spec or fixture identity mismatch")
    if sha(Path(sys.executable)) != spec["runtime"]["python"]["sha256"]:
        raise RuntimeError("Python runtime identity mismatch")
    if args.output.exists() or args.report.exists():
        raise RuntimeError("refusing to overwrite D11.1 Python quantizer output")

    adapter = json.loads(args.adapter_report.read_text())
    adapter_body = {key: value for key, value in adapter.items() if key != "reportHash"}
    if adapter.get("reportHash") != canonical_hash(adapter_body):
        raise RuntimeError("adapter report self-hash mismatch")
    if adapter.get("fixtureId") != args.fixture or adapter.get("repeat") != args.repeat:
        raise RuntimeError("adapter report cell mismatch")
    if adapter["arrays"]["motion"]["sha256"] != sha(args.input):
        raise RuntimeError("adapter motion binding mismatch")

    width, height = spec["scene"]["resolution"]
    payload = args.input.read_bytes()
    if len(payload) != width * height * 2 * 4:
        raise RuntimeError("motion input size mismatch")
    radius = float(spec["quantizerContract"]["acceptanceRadiusPixels"])
    encoded, maximum_error = quantize(payload, radius)

    args.output.parent.mkdir(parents=True, exist_ok=True)
    args.output.write_bytes(encoded)
    output_record = {
        "uri": str(args.output),
        "sha256": sha(args.output),
        "bytes": args.output.stat().st_size,
        "shape": [height, width, 2],
        "dtype": "little-endian-float32",
    }
    body = {
        "schemaVersion": "bfs.blenderNearestIntegerTemporalRecoveryPythonQuantizerReport.v0.1",
        "experimentId": spec["experimentId"],
        "fixtureId": args.fixture,
        "repeat": args.repeat,
        "producer": "python",
        "pid": os.getpid(),
        "runtime": {"python": sys.version.split()[0], "pythonExecutableSha256": sha(Path(sys.executable))},
        "adapterReport": {"uri": str(args.adapter_report), "sha256": sha(args.adapter_report)},
        "input": {"uri": str(args.input), "sha256": sha(args.input), "bytes": len(payload)},
        "output": output_record,
        "quantizer": {
            "candidate": "value >= 0 ? floor(value + 0.5) : ceil(value - 0.5)",
            "acceptanceRadiusPixels": radius,
            "wholeArrayAccepted": True,
            "positiveZeroCanonical": True,
            "idempotent": True,
        },
        "metrics": {"componentCount": len(payload) // 4, "maximumAbsoluteQuantizationErrorPixelsDecimal": f"{maximum_error:.18f}"},
        "operationCounts": {"pythonQuantizerProcesses": 1, "nodeQuantizerProcesses": 0, "modelCalls": 0, "networkCalls": 0},
    }
    args.report.parent.mkdir(parents=True, exist_ok=True)
    args.report.write_text(json.dumps({**body, "reportHash": canonical_hash(body)}, indent=2, sort_keys=True, allow_nan=False) + "\n")
    print(
        f"BFS_B52_D11_1_PYTHON_QUANTIZER_OK fixture={args.fixture} repeat={args.repeat} "
        f"components={len(payload) // 4} maxError={maximum_error:.12g} output={output_record['sha256']}"
    )


if __name__ == "__main__":
    main()

File: scripts/quantize-b52-d11-1-motion.mjs
#!/usr/bin/env node
// Bounded scalar-Node motion quantizer for B52-D11.1.

import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';

const SPEC_SHA256 = 'c4cb343672f53660d7c4ab69ccd489e00bb211e4aa1f489429f7a626ee48c42a';

function argumentsFrom(argv) {
  const values = {};
  for (let index = 0; index < argv.length; index += 2) values[argv[index]] = argv[index + 1];
  return values;
}

function sha(payload) {
  return crypto.createHash('sha256').update(payload).digest('hex');
}

function shaFile(filename) {
  return sha(fs.readFileSync(filename));
}

function stable(value) {
  if (Array.isArray(value)) return `[${value.map(stable).join(',')}]`;
  if (value && typeof value === 'object') {
    return `{${Object.keys(value).sort().map(key => `${JSON.stringify(key)}:${stable(value[key])}`).join(',')}}`;
  }
  return JSON.stringify(value);
}

function nearestInteger(value) {
  return value >= 0 ? Math.floor(value + 0.5) : Math.ceil(value - 0.5);
}

function quantize(payload, radius) {
  if (payload.length % 4 !== 0) throw new Error('motion payload is not a whole little-endian float32 array');
  const input = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
  const output = Buffer.alloc(payload.length);
  const outputView = new DataView(output.buffer, output.byteOffset, output.byteLength);
  let maximumError = 0;
  for (let index = 0; index < payload.length / 4; index += 1) {
    const value = input.getFloat32(index * 4, true);
    if (!Number.isFinite(value)) throw new Error(`QUANTIZER_DOMAIN nonfinite component=${index}`);
    const candidate = nearestInteger(value);
    const error = Math.abs(value - candidate);
    if (error > radius) {
      throw new Error(`QUANTIZER_DOMAIN component=${index} value=${value} candidate=${candidate} error=${error} radius=${radius}`);
    }
    maximumError = Math.max(maximumError, error);
    outputView.setFloat32(index * 4, candidate === 0 ? 0 : candidate, true);
  }
  const idempotent = quantizeIntegral(output);
  if (!idempotent.equals(output)) throw new Error('quantizer idempotence failure');
  return { output, maximumError };
}

function quantizeIntegral(payload) {
  const input = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
  const output = Buffer.alloc(payload.length);
  const outputView = new DataView(output.buffer, output.byteOffset, output.byteLength);
  for (let index = 0; index < payload.length / 4; index += 1) {
    const value = input.getFloat32(index * 4, true);
    const candidate = nearestInteger(value);
    outputView.setFloat32(index * 4, candidate === 0 ? 0 : candidate, true);
  }
  return output;
}

const args = argumentsFrom(process.argv.slice(2));
const required = ['--spec', '--fixture', '--repeat', '--input', '--adapter-report', '--output', '--report'];
if (required.some(key => !args[key])) throw new Error('missing required D11.1 quantizer argument');
const spec = JSON.parse(fs.readFileSync(args['--spec'], 'utf8'));
const fixture = spec.fixtures.find(item => item.id === args['--fixture']);
const repeat = Number(args['--repeat']);
if (shaFile(args['--spec']) !== SPEC_SHA256 || !fixture || ![1, 2].includes(repeat)) {
  throw new Error('B52-D11.1 spec, fixture or repeat identity mismatch');
}
if (shaFile(process.execPath) !== spec.runtime.node.sha256) throw new Error('Node runtime identity mismatch');
if (fs.existsSync(args['--output']) || fs.existsSync(args['--report'])) {
  throw new Error('refusing to overwrite D11.1 Node quantizer output');
}

const adapter = JSON.parse(fs.readFileSync(args['--adapter-report'], 'utf8'));
const adapterBody = Object.fromEntries(Object.entries(adapter).filter(([key]) => key !== 'reportHash'));
if (adapter.reportHash !== sha(Buffer.from(stable(adapterBody)))) throw new Error('adapter report self-hash mismatch');
if (adapter.fixtureId !== args['--fixture'] || adapter.repeat !== repeat) throw new Error('adapter report cell mismatch');
if (adapter.arrays.motion.sha256 !== shaFile(args['--input'])) throw new Error('adapter motion binding mismatch');

const [width, height] = spec.scene.resolution;
const payload = fs.readFileSync(args['--input']);
if (payload.length !== width * height * 2 * 4) throw new Error('motion input size mismatch');
const radius = spec.quantizerContract.acceptanceRadiusPixels;
const { output, maximumError } = quantize(payload, radius);

fs.mkdirSync(path.dirname(args['--output']), { recursive: true });
fs.writeFileSync(args['--output'], output);
const outputRecord = {
  uri: args['--output'], sha256: shaFile(args['--output']), bytes: output.length,
  shape: [height, width, 2], dtype: 'little-endian-float32',
};
const body = {
  schemaVersion: 'bfs.blenderNearestIntegerTemporalRecoveryNodeQuantizerReport.v0.1',
  experimentId: spec.experimentId,
  fixtureId: args['--fixture'],
  repeat,
  producer: 'node',
  pid: process.pid,
  runtime: { node: process.version, nodeExecutableSha256: shaFile(process.execPath) },
  adapterReport: { uri: args['--adapter-report'], sha256: shaFile(args['--adapter-report']) },
  input: { uri: args['--input'], sha256: shaFile(args['--input']), bytes: payload.length },
  output: outputRecord,
  quantizer: {
    candidate: 'value >= 0 ? floor(value + 0.5) : ceil(value - 0.5)',
    acceptanceRadiusPixels: radius,
    wholeArrayAccepted: true,
    positiveZeroCanonical: true,
    idempotent: true,
  },
  metrics: { componentCount: payload.length / 4, maximumAbsoluteQuantizationErrorPixelsDecimal: maximumError.toFixed(18) },
  operationCounts: { pythonQuantizerProcesses: 0, nodeQuantizerProcesses: 1, modelCalls: 0, networkCalls: 0 },
};
const report = { ...body, reportHash: sha(Buffer.from(stable(body))) };
fs.mkdirSync(path.dirname(args['--report']), { recursive: true });
fs.writeFileSync(args['--report'], `${JSON.stringify(report, null, 2)}\n`);
console.log(
  `BFS_B52_D11_1_NODE_QUANTIZER_OK fixture=${args['--fixture']} repeat=${repeat} `
  + `components=${payload.length / 4} maxError=${maximumError} output=${outputRecord.sha256}`,
);
